File: algorithms/medium/wiggle_subsequence.ex
defmodule Solution do
  @spec wiggle_max_length(nums :: [integer]) :: integer
  def wiggle_max_length([n | ns]) do
    ns
    |> Enum.reduce({1, 1, n}, fn n, {inc, dec, pre} ->
      cond do
        n > pre -> {inc, inc + 1, n}
        n < pre -> {dec + 1, dec, n}
        true -> {inc, dec, n}
      end
    end)
    |> then(fn {inc, dec, _} ->
      max(inc, dec)
    end)
  end
end

File: algorithms/medium/wiggle_subsequence.rs
impl Solution {
    pub fn wiggle_max_length(nums: Vec<i32>) -> i32 {
        let mut inc = 1;
        let mut dec = 1;
        for i in 1..nums.len() {
            if nums[i] > nums[i-1] {
                dec = inc + 1;
            } else if nums[i] < nums[i-1] {
                inc = dec + 1;
            } 
        }
        inc.max(dec)
    }
}
